waspc/src/Analyzer/TypeChecker/Monad.hs

```diff
@@ -0,0 +1,43 @@
+{-# LANGUAGE TupleSections #-}
+
+module Analyzer.TypeChecker.Monad where
+
+import Analyzer.Parser.AST
+import Analyzer.Type
+import Analyzer.TypeChecker.TypeError
+import qualified Analyzer.TypeDefinitions as TD
+import Control.Monad.Except (Except, runExcept, throwError)
+import Control.Monad.Reader (ReaderT, asks, runReaderT)
+import Control.Monad.State (StateT, evalStateT, gets, modify)
+import qualified Data.HashMap.Strict as H
+
+type Bindings = H.HashMap Ident Type
+
+type T a = StateT Bindings (ReaderT TD.TypeDefinitions (Except TypeError)) a
+
+lookupType :: Ident -> T (Maybe Type)
+lookupType ident = gets $ H.lookup ident
+
+setType :: Ident -> Type -> T ()
+setType ident typ = modify $ H.insert ident typ
+
+throw :: TypeError -> T a
```

Comment:
How is it that you implement this function instead of just letting people use `throwError`? I like this though, I think it makes it easier for others to know what they can do with this monad, but it is still a bit redundant. Therefore I am ok with doing this, just interested in the reasoning.

Reply:
True I forgot that! Cool, I like this as I said, it makes the "interface" more explicit.